File: MatlabBinaryFiles/readRWEDetectorData.m
function [data, error] = readRWEDetectorData(fid, fileInfo, data)
% reads binary data stored by the Right Whale Edge Detector.
%
% Inputs:
%   fid = file identifier
%   fileInfo = structure holding the file header, module header, a handle
%   data = a structure containing the standard data
%
% Output:
%   data = structure containing data from a single object
%

% initialize variables
error=false;

try
    
    % read module specific data
    dataLength = fread(fid, 1, 'int32');
    if (dataLength==0)
        return;
    end

    if (fileInfo.moduleHeader.version==0)
        data.startSample = fread(fid, 1, 'int64');
        data.channelMap = fread(fid, 1, 'int32');
    end
    
    data.type = fread(fid, 1, 'int16');
    data.signal = fread(fid, 1, 'float');
    data.noise = fread(fid, 1, 'float');
    data.nSlices = fread(fid, 1, 'int16');
    
    data.times = zeros(1, data.nSlices);
    data.sliceNums = zeros(1, data.nSlices);
    data.loFreqs = zeros(1, data.nSlices);
    data.peakFreqs = zeros(1, data.nSlices);
    data.hiFreqs = zeros(1, data.nSlices);
    data.peakAmps = zeros(1, data.nSlices);
    for i = 1:data.nSlices
        data.sliceNums(i) = fread(fid, 1, 'int16');
        data.loFreqs(i) = fread(fid, 1, 'int16');
        data.peakFreqs(i) = fread(fid, 1, 'int16');
        data.hiFreqs(i) = fread(fid, 1, 'int16');
        data.peakAmps(i) = fread(fid, 1, 'float');
    end
    
catch mError
    disp(['Error reading ' fileInfo.fileHeader.moduleType '  data object.  Data read:']);
    disp(data);
    disp(getReport(mError));
    error=true;
end
    

File: HelperFuncs/readRWEDetectorData.R
# Reads binary data stored by the Right Whale Edge Detector
#
# Inputs:
#     fid = file identifier
#     fileInfo = structure holding the file header, module header, a handle
#     data = a structure containing the standard data
# Output:
#     data = structure containing data from a single object

readRWEDetectorData <- function(fid, fileInfo, data) {
    error <- FALSE
    
    tryCatch({
        dataLength <- pamBinRead(fid, 'int32', n=1)
        if(dataLength==0) {
            return(list(data=data, error=error))
        }
        
        version <- fileInfo$moduleHeader$version
        
        if(version==0) {
            data$startSample <- pamBinRead(fid, 'int64', n=1)
            data$channelMap <- pamBinRead(fid, 'int32', n=1)
        }
        
        data$type <- pamBinRead(fid, 'int16', n=1)
        data$signal <- pamBinRead(fid, 'float', n=1)
        data$noise <- pamBinRead(fid, 'float', n=1)
        data$nSlices <- pamBinRead(fid, 'int16', n=1)
        
        zeros <- rep(0, data$nSlices)
        
        data$times <- zeros
        data$sliceNums <- zeros
        data$loFreqs <- zeros
        data$peakFreqs <- zeros
        data$hiFreqs <- zeros
        data$peakAmps <- zeros
        
        for(i in 1:data$nSlices) {
            data$sliceNums[i] <- pamBinRead(fid, 'int16', n=1)
            data$loFreqs[i] <- pamBinRead(fid, 'int16', n=1)
            data$peakFreqs[i] <- pamBinRead(fid, 'int16', n=1)
            data$hiFreqs[i] <- pamBinRead(fid, 'int16', n=1)
            data$peakAmps[i] <- pamBinRead(fid, 'float', n=1)
        }
        
        return(list(data=data, error=error))
    }, warning = function(w) {
        print(paste('Warning occurred: ', w))
        return(list(data=data, error=error))
    }, error = function(e) {
        print(paste('Error reading ', fileInfo$fileHeader$moduleType, ' data object. Data read:'))
        print(data)
        print(e)
        error <- TRUE
        return(list(data=data, error=error))
    })
}